crates/ra_assists/src/handlers/replace_let_with_if_let.rs

```diff
@@ -0,0 +1,108 @@
+use hir::Adt;
+use ra_syntax::{
+    ast::{self, make},
+    AstNode, T,
+};
+
+use crate::{
+    assist_ctx::{Assist, AssistCtx},
+    AssistId,
+};
+use ast::edit::{AstNodeEdit, IndentLevel};
+use std::iter::once;
+
+// Assist: replace_let_with_if_let
+//
+// Replaces `if let` with an else branch with a `match` expression.
+//
+// ```
+// # enum Option<T> { Some(T), None }
+//
+// fn main(action: Action) {
+//     <|>let x = compute();
+// }
+//
+// fn compute() -> Option<i32> { None }
+// ```
+// ->
+// ```
+// # enum Option<T> { Some(T), None }
+//
+// fn main(action: Action) {
+//     if let Some(x) = compute() {
+//     }
+// }
+//
+// fn compute() -> Option<i32> { None }
+// ```
+pub(crate) fn replace_let_with_if_let(ctx: AssistCtx) -> Option<Assist> {
+    let let_kw = ctx.find_token_at_offset(T![let])?;
+    let let_stmt = let_kw.ancestors().find_map(ast::LetStmt::cast)?;
+    let init = let_stmt.initializer()?;
+    let original_pat = let_stmt.pat()?;
+    let ty = ctx.sema.type_of_expr(&init)?;
+    let enum_ = match ty.as_adt() {
+        Some(Adt::Enum(it)) => it,
+        _ => return None,
+    };
+    let happy_case =
+        [("Result", "Ok"), ("Option", "Some")].iter().find_map(|(known_type, happy_case)| {
```

Comment:
maybe we should centralize this knowledge somewhere, since there are now a few assists doing related things

Reply:
As these are assists, I think it makes more sense to wait for a while more, until a more clear usage pattern emerges. Assists are very easy to refactor and are super-isolated, so copy-paste is not really a problem.